components/script/dom/event.rs

```diff
@@ -147,49 +166,195 @@ impl Event {
+            target
+                .downcast::<Element>()
+                .and_then(|e| e.as_maybe_activatable())
+        } else {
+            // Step 3
+            None
+        };
+
+        // Steps 5-6 - 5.7 are shadow DOM slot things
+
+        // Step 5.9.8.1, not covered in construct_event_path
+        // This what makes sure that clicking on e.g. an <img> inside
+        // an <a> will cause activation of the activatable ancestor.
+        if is_activation_event && activation_target.is_none() && self.bubbles.get() {
+            for object in event_path.iter() {
+                if let Some(activatable_ancestor) = object
+                    .downcast::<Element>()
+                    .and_then(|e| e.as_maybe_activatable())
+                {
+                    activation_target = Some(activatable_ancestor);
+                    // once activation_target isn't null, we stop
+                    // looking at ancestors for it.
+                    break;
+                }
+            }
+        }
+
+        // Steps 5.10-5.11 are shadow DOM
+
+        // Not specified in dispatch spec overtly; this is because
+        // the legacy canceled activation behavior of a checkbox
+        // or radio button needs to know what happened in the
+        // corresponding pre-activation behavior.
+        let mut pre_activation_result: Option<InputActivationState> = None;
+
+        // Step 5.12
+        if is_activation_event {
+            if let Some(maybe_checkbox) = activation_target {
+                pre_activation_result = maybe_checkbox.legacy_pre_activation_behavior();
+            }
+        }
+
+        let timeline_window = match DomRoot::downcast::<Window>(target.global()) {
+            Some(window) => {
+                if window.need_emit_timeline_marker(TimelineMarkerType::DOMEvent) {
+                    Some(window)
+                } else {
+                    None
+                }
+            },
+            _ => None,
+        };
+
+        // Step 5.13
+        for object in event_path.iter().rev() {
+            if &**object == &*target {
+                self.phase.set(EventPhase::AtTarget);
+            } else {
+                self.phase.set(EventPhase::Capturing);
+            }
+
+            // setting self.target is step 1 of invoke,
+            // but done here because our event_path isn't a member of self
+            // (without shadow DOM, target_override is always the
+            // target to set to)
+            self.target.set(Some(target_override));
+            invoke(
+                timeline_window.as_deref(),
+                object,
+                self,
+                Some(ListenerPhase::Capturing),
+            );
+        }
+
+        // Step 5.14
+        for object in event_path.iter() {
+            if &**object == &*target {
+                self.phase.set(EventPhase::AtTarget);
+
+                self.target.set(Some(target_override));
```

Comment:
Seems like we could de-duplicate these two blocks by only setting phase inside the if/else.

Reply:
Yes, that would be easier to read. Thank you.